### PR TITLE
Schema: harmonize mobility messages in 2.0.0 versions

### PR DESCRIPTION
**Pull request by Hugues360**
Fixes: #137

### CODE REVIEW

**Comment by nbuffon** [comment]
For now SDK and application do not allow to use a specific version of a schema, and then are considered using the last version by default  
Changes like renaming a field (see commit d4ccdee0 for example) would make the use of this new version (the would then be last one) incompatible with the implementation, like the Rust one would no longer be able to deserialize any incoming message  
Introducing such changes should also provide the changes in the implementation in each languages  

@ymorin-orange @mathieu1fb any opinion on this ?

File: schema/cam/cam_schema_2-0-0.json [hunk 1]
```diff
@@ -31,7 +31,7 @@
     "version": {
       "type": "string",
       "description": "json message format version",
-      "const": "1.99.99"
+      "const": "2.0.0"
```
[the remark this answers is not in the document]
Comment (Hugues360) [reply schema/cam/cam_schema_2-0-0.json:34]: In fact, seen with Yann, it's a way to have only the right version number (2.0.0 in this case) in the very last commit including all the updates. If someone pull an "intermediate" commit, before the last one, he won't have the right version in the schema but a version with .99 indicating it's a developing version.

**Comment by ymorin-orange** [comment]
> For now SDK and application do not allow to use a specific version of a schema, and then are considered using the last version by default Changes like renaming a field (see commit [d4ccdee](https://github.com/Orange-OpenSource/its-client/commit/d4ccdee00055273b1f57a4adec61f7eda6b5c550) for example) would make the use of this new version (the would then be last one) incompatible with the implementation, like the Rust one would no longer be able to deserialize any incoming message Introducing such changes should also provide the changes in the implementation in each languages
> 
> @ymorin-orange @mathieu1fb any opinion on this ?

1. First and foremost, the commit log should explain the reasoning for this renaming, if not to avoid such comment, but at the very least for reference for the future, when we all have long forgotten about this.
2. Our applications are for now only emitting a single version (the current-latest one) of each schema; just updating the schemas would not magically make all those applications emit the new versions, and thus no message of a newer version would be received by applications. So it is totally acceptable to create new schemas with an updated version in which we chagne the fields; that is explicitly the reason why we do have a version in them.
3. If an implementation can't cope with versions in does not recognise, but just ignores those mesages, that's perfectly acceptable too, and is again the reason for the versioning. If however an implementation crashes, then that's a bug in that implementation, and should be fixed (to at least ignore unknown messages, or ignore versions of known messages); indeed, once we start updating an application, then newer messages will be emitted, and that will have to be handled (in any suitable way: ignore or use) by applications. Note the fifth point below, too.
4. Updating applications to use those newer schemas should be scheduled as son as the newer schemas are validated.
5. Currently, we do not validate the messages against the schmas before we use those messages, which opens up the door to problems in interpreting the messages, and this is the crux of the issue. When we add validation to our applications (via the SDK), then it will be easier to have a mixed-versions interchange of schemas, where applications would just ignore messages that do not match any schema (and thus any version of such schema) they know of.

(Note: I would be surprised if our python apps would have no issue with newer schemas as well... Although they mostly creates them and do not pasrse them for now, so the issue would be largely mitigated already (Famous Last Words™...)

File: schema/information/information_schema_2-0-0.json [hunk 2]
```diff
@@ -26,7 +26,7 @@
       "description": "json message format version",
       "const": "1.99.99"
     },
-    "instance_id": {
+    "source_uuid": {
       "description": "unique id all other the world for a server",
```
Comment (ymorin-orange) [inline schema/information/information_schema_2-0-0.json:30]: Could you also fix this error (note: it exists in many schemas): s/all other/all over/, please?

File: schema/cam/cam_schema_2-0-0.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (ymorin-orange) [reply schema/cam/cam_schema_2-0-0.json:34]: Indeed, this avoids cases where two files, at two different revisions, have the same version:
1. if we add/delete/rename fields while still keeping the old version number, then we have multiple schemas with the old version, and it is not easy to find which is the correct one;
2. if we first change the version, then the same is true, but with the new version.

We can't really rely on the filenames, because filenames can change; we can only rely on the version in the schemas themselves. Thus, it is important to denote that a schema is not stable, and that it should not be used; this is the role of the _.99_ version, which clearly inidicates that.

When discussing this with @Hugues360, I subtly hinted that we should have documentation that explains our process at maintaining those schmas. Way too subtly, maybe... ;-) More generally, though, we should have a documentation that explains the versioning scheme for each component (java, python, rust, schemas).

File: schema/information/information_schema_2-0-0.json (hunk 2 above)
Comment (Hugues360) [reply schema/information/information_schema_2-0-0.json]: Ok, fixed

Comment (mathieu1fb) [reply schema/information/information_schema_2-0-0.json]: I don't think that Yann meant to replace `instance_id` by `source_uuid` as they do not represent the same information. He was referring to the spelling mistakes in the description.

Comment (ymorin-orange) [reply schema/information/information_schema_2-0-0.json]: > I don't think that Yann meant to replace `instance_id` by `source_uuid` as they do not represent the same information. He was referring to the spelling mistakes in the description.

1. Indeed, I was speaking about the typo, and Hugues did fix it in the new iteration.
2. We still want to harmonise the top-level fields so that all messages still contain the same "header":
    * `message_type`: the type of message this is
    * `version`: the version of the schema that applies
    * `timestamp`: the UNIX timestamp the message was created at
    * `source_uuid`: the universally unique ID of the source that emitted the message
    * `origin`: the (type of) entity that emitted the message
    * `message`: the JSON-translation of the ETSI message

So far, only the schema for _information_ message defines an `instance_id`. If that is not the same thing as `source_uuid`, then we indeed must **add** `source_uuid` and keep `instance_id`.

Note that CAM, for example, has a `message.station_id`. Is that the same thing as the `instance_id` from the information message, or is it yet something else?

**Comment by ymorin-orange** [comment]
@Hugues360 I'm OK with the cahnges, good work!

However, I have some comments about the messages in the commit logs, but unfortunately GitHub does not allow commenting on commit logs... So, here are my concerns on a few of those commits:

* 0ca7bab schema/ : Changing field 'type' into 'message_type'
    * it is obvious when looking at the diff that the field is being renamed. What should be present in the commit log is the reason why it is being renamed, so that when in the future we are looking back at that commit, we do not have to wonder why we did that rename, or that we do not have to chase documentation that might have disapeared then...
* 589bf7f schema: deleting origin field in schema having this field
    * ditto: the commit log should explain why we remove that field
* 02ceb3a schema/status/status_schema_2-0-0.json: cellular and gnss are no longer required fields
    * ditto: the commit log should explain why those fields are no longer mandatory

Basically, a commit log should _describe_ the change; it should _explain_ it, and provide the _rationale_ for the change. Of course, there are commits where the commit log would provide no additional value (e.g. the fix for the wording error does not need to be explained, it is so obvious). But when the "why" can't be answered just by looking at the commit, then the commit log is missing information.

Also, the title of the commit log should be imperative, not past tense. For example, it should be something like:

    * `schemas: change field 'type' to 'message_type'`
    * `schemas: remove unused field 'origin'`
    * `schema/information: fix wording error`

and so on...

**Comment by Hugues360** [comment]
> @Hugues360 I'm OK with the changes, good work!
> 
> However, I have some comments about the messages in the commit logs, but unfortunately GitHub does not allow commenting on commit logs... So, here are my concerns on a few of those commits:
> 
> * [0ca7bab](https://github.com/Orange-OpenSource/its-client/commit/0ca7bab8dc5dfb644985378fc3948b739dc74c26) schema/ : Changing field 'type' into 'message_type'
>   
>   * it is obvious when looking at the diff that the field is being renamed. What should be present in the commit log is the reason why it is being renamed, so that when in the future we are looking back at that commit, we do not have to wonder why we did that rename, or that we do not have to chase documentation that might have disapeared then...
> * [589bf7f](https://github.com/Orange-OpenSource/its-client/commit/589bf7f365ce79d8d94dcef288a683ff0d759bdd) schema: deleting origin field in schema having this field
>   
>   * ditto: the commit log should explain why we remove that field
> * [02ceb3a](https://github.com/Orange-OpenSource/its-client/commit/02ceb3ae59e93e89e01ed87284d90d71c337c740) schema/status/status_schema_2-0-0.json: cellular and gnss are no longer required fields
>   
>   * ditto: the commit log should explain why those fields are no longer mandatory
> 
> Basically, a commit log should _describe_ the change; it should _explain_ it, and provide the _rationale_ for the change. Of course, there are commits where the commit log would provide no additional value (e.g. the fix for the wording error does not need to be explained, it is so obvious). But when the "why" can't be answered just by looking at the commit, then the commit log is missing information.
> 
> Also, the title of the commit log should be imperative, not past tense. For example, it should be something like:
> 
> ```
> * `schemas: change field 'type' to 'message_type'`
> * `schemas: remove unused field 'origin'`
> * `schema/information: fix wording error`
> ```
> 
> and so on...

Thank you Yann for the comments.
I have updated the commit messages

**Review by ymorin-orange — APPROVED** [review]

**Comment by ymorin-orange** [comment]
@nbuffon @mathieu1fb Last chance for a comment of your own! Hurry up if you still want to chime in! ;-)
